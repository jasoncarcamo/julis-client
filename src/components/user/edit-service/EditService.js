import React from 'react';
import {withRouter} from 'react-router-dom';
import tokenService from '../../../services/TokenService';
import queryString from 'query-string';
import DatePicker from 'react-date-picker';
import TimePicker from 'react-time-picker';
import './editservice.css';

class EditService extends React.Component{
    constructor(props){
        super(props);
        this.state = {
            services: [
                {
                    name: 'Windows',
                    price: '20.00'
                },
                {
                    name: 'Refridgerator',
                    price: '20.00'
                },
                {
                    name: 'Walls',
                    price: '30.00'
                }
            ],
            service_type: '',
            comments: '',
            time: '',            
            date: new Date(),
            price: '0.00'
        }
    }

    componentDidMount(){
        
        const serviceId = queryString.parse(this.props.location.search);
        fetch(`https://fathomless-eyrie-65525.herokuapp.com/user/service?id=${serviceId.id}`, {
            headers: {
                "content-type": "application/json",
                "authorization": `bearer ${tokenService.getAuthToken()}`
            }
        })
        .then(res => res.json())
        .then(resData => {
            console.log(resData)
            this.handleEditService(resData.services, serviceId.id);
        });
    }

    handleEditService = (services, serviceId) =>{
        const inputArray = document.getElementsByTagName("input");
        const currentService = services.find(service => service.id === Number(serviceId));
        let serviceArray;
        if(currentService.service_type.includes(",")){
            serviceArray = currentService.service_type.split(',');
            for(let i = 0; i < serviceArray.length; i++){
                inputArray[serviceArray[i]].click()
                console.log(inputArray[serviceArray[i]])
            }
            console.log(serviceArray)
        } else{
            console.log('Bye')
        }
    }

    handleService= (e)=>{
        let newService;
        let newPrice;
        let serviceArray;
        let serviceIndex;
        let targetName;

        if(e.target.checked){
            
            newService = this.state.service_type + ',' + e.target.name;     
            serviceArray = newService.split(',');      
            
            if(serviceArray[0] === ""){
                serviceArray.shift();                
            }
            
            newService = serviceArray.join(',');
            newPrice = parseFloat(this.state.price) + parseFloat(e.target.value)
                        
        } else{
            
            targetName = e.target.name
            serviceArray = this.state.service_type.split(',');
            serviceIndex = serviceArray.indexOf(targetName)
            
            serviceArray.splice(serviceIndex, 1)
            if(serviceArray[0] === ""){
                serviceArray.shift()
            }
            
            newService = serviceArray.join(', ')
            
            newPrice = parseFloat(this.state.price) - parseFloat(e.target.value)
        }
        
        this.setState({service_type: newService})
        this.setState({ price: newPrice});
       
    }

    handleComments = (e)=>{
        
        this.setState({ comments: e.target.value})
    }

    handleDay = (date)=>{
        this.setState({date})
    }

    handleTime = (time)=>{
        this.setState({time})
    }

    formatTime = ()=>{
        let newTime;
        const arr = this.state.time.split(':');
        if(arr[0] > 12){
            arr[0] = arr[0] - 12 ;
            newTime = arr.join(':') + ' PM';
        } else if(arr[0] === '00'){
            arr[0] = '12'
            newTime = arr.join(':') + ' AM'
        } else if(arr[0] === '12'){
            newTime = arr.join(':') + ' PM'
        }
        else{
            newTime = arr.join(':') + ' AM';
        }
        return newTime;
    }
    
    
    handlePrice = (e)=>{
        this.setState({ price: e.target.value})
    }

    handleSubmit = (e)=>{
        
        e.preventDefault();
        const serviceId = queryString.parse(this.props.location.search);
        fetch(`https://fathomless-eyrie-65525.herokuapp.com/user/service`, {
            method: 'PATCH',
            headers: {
                'content-type': 'application/json',
                'authorization': `bearer ${tokenService.getAuthToken()}`
            },
            body: JSON.stringify({service_type: this.state.service_type, comments: this.state.comments, day: this.state.date, best_time_reached: `${this.formatTime()}`, price: this.state.price, id: serviceId.id, date_modified: new Date()})
        })
        this.props.history.push('/user/services')
    }

    render(){

        return (
            <section id="edit_section">
                <form onSubmit={this.handleSubmit} id="edit_form">
                    <fieldset>
                        <header><h1>Edit your service.</h1></header>

                        <div className="checkbox-grid">
                            <div>
                                {this.state.services.map( service => (
                                    <label key={service.name} htmlFor={`edit_${service.name}`}>
                                        <input type="checkbox" id={`edit_${service.name}`} name={service.name} value={service.price} onChange={this.handleService}>                               
                                        </input>{service.name}
                                    </label>)
                                )}
                            </div>

                            <span id="edit_price">${this.state.price}</span>
                        </div>

                        <label htmlFor="edit_day">When?</label>
                        <DatePicker className="edit_days" value={this.state.date} onChange={this.handleDay} required/>

                        <label htmlFor="edit_time">Time:
                        </label>
                        <TimePicker id="edit_time" onChange={this.handleTime} value={this.state.time} disableClock={true} />

                        <label htmlFor="edit_comments">Comments
                        ?</label>
                        <textarea id="edit_comments" onChange={this.handleComments} value={this.state.comments} placeholder="Any special request?"></textarea>

                        <button type="submit" id="edit_submit">Go</button>
                    </fieldset>
                </form>
            </section>
        )
    }
}

export default withRouter(EditService);