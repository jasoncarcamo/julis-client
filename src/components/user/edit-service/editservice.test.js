import React from 'react';
import ReactDOM from 'react-router-dom';
import {BrowserRouter} from 'react-router-dom';
import Enzyme, {shallow} from 'enzyme';
import renderer from 'react-test-renderer';
import EditService from './EditService';
import Adaptor from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adaptor()});

describe('<EditService/>', ()=>{
    let wrapper;

    beforeEach(()=>{

        wrapper = shallow(<EditService/>);

        
    })

    it('create snapshot for EditService component', ()=>{
        const tree = renderer.create(<BrowserRouter><EditService/></BrowserRouter>).toJSON();

        expect(tree).toMatchSnapshot();
    })

    it('has submit button', ()=>{

        expect(wrapper.containsAnyMatchingElements([
            <button
            id="edit_submit"
            type="submit"
            >
                Go
            </button>
        ]))
    });

    it('simulates click on submit button', ()=>{
        
        wrapper = shallow(<EditService onClick={()=> console.log('Click!')}></EditService>)

        wrapper.simulate('click');
    })
})