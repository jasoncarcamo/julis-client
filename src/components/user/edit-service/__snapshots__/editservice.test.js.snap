// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditService/> create snapshot for EditService component 1`] = `
<section
  id="edit_section"
>
  <form
    id="edit_form"
    onSubmit={[Function]}
  >
    <fieldset>
      <header>
        <h1>
          Edit your service.
        </h1>
      </header>
      <div
        className="checkbox-grid"
      >
        <div>
          <label
            htmlFor="edit_Windows"
          >
            <input
              id="edit_Windows"
              name="Windows"
              onChange={[Function]}
              type="checkbox"
              value="20.00"
            />
            Windows
          </label>
          <label
            htmlFor="edit_Refrigerator"
          >
            <input
              id="edit_Refrigerator"
              name="Refrigerator"
              onChange={[Function]}
              type="checkbox"
              value="20.00"
            />
            Refrigerator
          </label>
          <label
            htmlFor="edit_Walls"
          >
            <input
              id="edit_Walls"
              name="Walls"
              onChange={[Function]}
              type="checkbox"
              value="30.00"
            />
            Walls
          </label>
        </div>
        <span
          id="edit_price"
        >
          $
          0.00
        </span>
      </div>
      <label
        htmlFor="edit_day"
      >
        When?
      </label>
      <div
        className="react-date-picker react-date-picker--closed react-date-picker--enabled edit_days"
        onFocus={[Function]}
      >
        <div
          className="react-date-picker__wrapper"
        >
          <div
            className="react-date-picker__inputGroup"
            onClick={[Function]}
            role="presentation"
          >
            <input
              max="275760-09-12"
              min="-271821-04-19"
              name="date"
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              style={
                Object {
                  "left": "-9999px",
                  "position": "absolute",
                  "top": "-9999px",
                  "visibility": "hidden",
                }
              }
              type="date"
              value="2019-09-28"
            />
            
            <input
              autoComplete="off"
              className="react-date-picker__inputGroup__input react-date-picker__inputGroup__month"
              max={12}
              min={1}
              name="month"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              placeholder="--"
              required={true}
              type="number"
              value={9}
            />
            <span
              className="react-date-picker__inputGroup__divider"
            >
              /
            </span>
            <input
              autoComplete="off"
              className="react-date-picker__inputGroup__input react-date-picker__inputGroup__day"
              max={30}
              min={1}
              name="day"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              placeholder="--"
              required={true}
              type="number"
              value={28}
            />
            <span
              className="react-date-picker__inputGroup__divider"
            >
              /
            </span>
            <input
              autoComplete="off"
              className="react-date-picker__inputGroup__input react-date-picker__inputGroup__year"
              max={275760}
              min={1000}
              name="year"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              placeholder="----"
              required={true}
              step={1}
              type="number"
              value={2019}
            />
            
          </div>
          <button
            className="react-date-picker__clear-button react-date-picker__button"
            onClick={[Function]}
            onFocus={[Function]}
            type="button"
          >
            <svg
              className="react-date-picker__clear-button__icon react-date-picker__button__icon"
              height={19}
              stroke="black"
              strokeWidth={2}
              viewBox="0 0 19 19"
              width={19}
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="4"
                x2="15"
                y1="4"
                y2="15"
              />
              <line
                x1="15"
                x2="4"
                y1="4"
                y2="15"
              />
            </svg>
          </button>
          <button
            className="react-date-picker__calendar-button react-date-picker__button"
            onClick={[Function]}
            onFocus={[Function]}
            type="button"
          >
            <svg
              className="react-date-picker__calendar-button__icon react-date-picker__button__icon"
              height={19}
              stroke="black"
              strokeWidth={2}
              viewBox="0 0 19 19"
              width={19}
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="none"
                height="15"
                width="15"
                x="2"
                y="2"
              />
              <line
                x1="6"
                x2="6"
                y1="0"
                y2="4"
              />
              <line
                x1="13"
                x2="13"
                y1="0"
                y2="4"
              />
            </svg>
          </button>
        </div>
      </div>
      <label
        htmlFor="edit_time"
      >
        Time:
      </label>
      <div
        className="react-time-picker react-time-picker--closed react-time-picker--enabled"
        onFocus={[Function]}
      >
        <div
          className="react-time-picker__wrapper"
        >
          <div
            className="react-time-picker__inputGroup"
            onClick={[Function]}
            role="presentation"
          >
            <input
              max={null}
              min={null}
              name="time"
              onChange={[Function]}
              onFocus={[Function]}
              step={60}
              style={
                Object {
                  "left": "-9999px",
                  "position": "absolute",
                  "top": "-9999px",
                  "visibility": "hidden",
                }
              }
              type="time"
              value=""
            />
            
            <input
              autoComplete="off"
              className="react-time-picker__inputGroup__input react-time-picker__inputGroup__hour"
              max={12}
              min={1}
              name="hour12"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              placeholder="--"
              required={null}
              type="number"
              value=""
            />
            <span
              className="react-time-picker__inputGroup__divider"
            >
              :
            </span>
            <input
              autoComplete="off"
              className="react-time-picker__inputGroup__input react-time-picker__inputGroup__minute"
              max={59}
              min={0}
              name="minute"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              placeholder="--"
              required={null}
              type="number"
              value=""
            />
            <span
              className="react-time-picker__inputGroup__divider"
            >
               
            </span>
            <select
              className="react-time-picker__inputGroup__input react-time-picker__inputGroup__amPm"
              name="amPm"
              onChange={[Function]}
              required={null}
              value=""
            >
              <option
                value=""
              >
                --
              </option>
              <option
                value="am"
              >
                AM
              </option>
              <option
                value="pm"
              >
                PM
              </option>
            </select>
            
          </div>
          <button
            className="react-time-picker__clear-button react-time-picker__button"
            onClick={[Function]}
            onFocus={[Function]}
            type="button"
          >
            <svg
              className="react-time-picker__clear-button__icon react-time-picker__button__icon"
              height={19}
              stroke="black"
              strokeWidth={2}
              viewBox="0 0 19 19"
              width={19}
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="4"
                x2="15"
                y1="4"
                y2="15"
              />
              <line
                x1="15"
                x2="4"
                y1="4"
                y2="15"
              />
            </svg>
          </button>
        </div>
      </div>
      <label
        htmlFor="edit_comments"
      >
        Comments ?
      </label>
      <textarea
        id="edit_comments"
        onChange={[Function]}
        placeholder="Any special request?"
        value=""
      />
      <button
        id="edit_submit"
        type="submit"
      >
        Go
      </button>
    </fieldset>
  </form>
</section>
`;
